libfprint/drivers/goodixtls/goodix5395.c

```diff
@@ -332,6 +332,14 @@ static void fpi_device_goodixtls5395_write_psk(FpDevice *dev, FpiSsm *ssm) {
     }
 }
 
+void fpi_device_goodixtls5395_setup_finger_down_detection(FpDevice *dev, FpiSsm *ssm){
+    GError *error = NULL;
+    if(!fpi_goodix_device_gtls_connection(dev, error)){
+        FAIL_SSM_WITH_RETURN(ssm, error)
+        fp_dbg("Error!");
```

Comment:
it will not be executed, Should before FAIL_SSM_WITH_RETURN

look at FAIL_SSM_WITH_RETURN define

Reply:
The idea here is fail ssm (sequence state machine) (You can look at FAIL_SSM_WITH_RETURN define it is failed current ssm and return)

I try to handle error in the next methods: ping, check_firmware and etc. All methods called by these methods that may fail must return gboolean and accept the GError** parameter. 
If an error occurs, the method must set an error and return FALSE.

The higher level methods should check the return value and if it is false, they should fail ssm and pass this error.

In case when you check receive data and this is not valid. You should fail ssm, create GError* object and pass it to FAIL_SSM_WITH_RETURN.


I think we should refactor the code and unify the error handling functionality, but not now.